language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -421,10 +530,155 @@ genModuleRef (pkgRef, modName) = case pkgRef of
+  , "    \"@typescript-eslint/no-inferrable-types\": \"off\""
+  , "  }"
+  , "}"
+  ]
+
+writePackageJson :: FilePath -> String -> String -> Scope -> [Dependency] -> IO ()
+writePackageJson packageDir damlTypesVersion packageVersion (Scope scope) depends =
+  writeFileUTF8 (packageDir </> "package.json") $ unlines
+  (["{"
+   , "  \"private\": true,"
+   , "  \"name\": \"" <> packageName <> "\","
+   , "  \"version\": \"" <> packageVersion <> "\","
+   , "  \"description\": \"Produced by daml2ts\","
+   , "  \"license\": \"Apache-2.0\","
+   , "  \"dependencies\": {"
+   , "    \"@daml/types\": \"" <> damlTypesVersion <> "\","
+   , "    \"@mojotech/json-type-validation\": \"^3.1.0\"" ++ if not $ null dependencies then ", " else ""
+   ] ++ dependencies ++
+    ["  },"
+    , "  \"scripts\": {"
+    , "    \"build\": \"tsc --build\","
+    , "    \"lint\": \"eslint --ext .ts src/ --max-warnings 0\""
+    , "  },"
+    , "  \"devDependencies\": {"
+    , "    \"@typescript-eslint/eslint-plugin\": \"^2.11.0\","
+    , "    \"@typescript-eslint/parser\": \"^2.11.0\","
+    , "    \"eslint\": \"^6.7.2\","
+    , "    \"typescript\": \"~3.7.3\""
+    , "  }"
+    , "}"
+    ])
+  where
+    packageName =  packageNameOfPackageDir packageDir
+    dependencies = withCommas [ "    \"" <> pkg <> "\": \"" <> packageVersion <> "\""
+                              | d <- depends
+                              , let pkg = scope ++ "/" ++ undependency d
+                              ]
+
+    withCommas :: [String] -> [String]
+    withCommas [] = []
+    withCommas ms = reverse (head ms' : map (++",") (tail ms')) where ms' = reverse ms
+
+-- This type describes the format of a "top-level" 'package.json'. We
+-- expect such files to have the format
+-- {
+--   "private": true,
+--   "workspaces: [
+--      "path/to/foo",
+--      "path/to/bar",
+--      ...
+--   ]
+-- }
+data PackageJson = PackageJson
+  {  private :: Bool
+  ,  workspaces :: [T.Text]
+  } deriving Show
+-- Explicitly provide instances to avoid relying on restricted
+-- extension 'DeriveGeneric'.
+instance FromJSON PackageJson where
+  parseJSON (Object v) = PackageJson
+      <$> v .: "private"
+      <*> v .: "workspaces"
+  parseJSON _ = mzero
+instance ToJSON PackageJson where
+  toJSON (PackageJson private workspaces) =
+    object [ "private" .= private
+           , "workspaces" .= workspaces
+           ]
+
+-- Read the provided 'package.json'; transform it to include the
+-- provided workspaces; write it back to disk.
+writeTopLevelPackageJson :: FilePath -> [(String, String, [Dependency])] -> FilePath -> IO ()
+writeTopLevelPackageJson optOutputDir dependencies file = do
+  let (g, nodeFromVertex) = graphFromEdges'
+        (map (\(a, b, ds) -> (a, b, map undependency ds)) dependencies)
+      ps = map (fst3 . nodeFromVertex) $ reverse (topSort g)
+        -- Topologically order our packages.
+      ldr = fromJust (stripPrefix "@" (unscope (scopeOfScopeDir optOutputDir)))
```

Comment:
Let’s leave it for a separate PR but the `stripPrefix` stuff here looks similar to the other `stripPrefix` stuff. Rather than stripping it here, it shouldn’t be added in the first place.

Reply:
i'll work on a follow up for this today.

Reply:
https://github.com/digital-asset/daml/pull/4818